File: src/git.ts
import { spawnSync } from "node:child_process";
import { basename } from "node:path";

let cachedProject: string | undefined;
let cachedBranch: string | undefined;
let resolved = false;

function resolveGit(): void {
	if (resolved) return;
	resolved = true;

	try {
		const toplevel = spawnSync("git", ["rev-parse", "--show-toplevel"], {
			encoding: "utf-8",
			timeout: 3000,
		});
		if (toplevel.status === 0 && toplevel.stdout) {
			cachedProject = basename(toplevel.stdout.trim());
		}
	} catch {}

	try {
		const branch = spawnSync("git", ["rev-parse", "--abbrev-ref", "HEAD"], {
			encoding: "utf-8",
			timeout: 3000,
		});
		if (branch.status === 0 && branch.stdout) {
			cachedBranch = branch.stdout.trim();
		}
	} catch {}
}

export function inferGitProject(): string | undefined {
	if (process.env.RELOG_PROJECT) return process.env.RELOG_PROJECT;
	resolveGit();
	return cachedProject;
}

export function inferGitBranch(): string | undefined {
	if (process.env.RELOG_BRANCH) return process.env.RELOG_BRANCH;
	resolveGit();
	return cachedBranch;
}

File: python/src/relog/_git.py
from __future__ import annotations

import os
import subprocess

_cached_project: str | None = None
_cached_branch: str | None = None
_resolved = False


def _resolve_git() -> None:
    global _cached_project, _cached_branch, _resolved
    if _resolved:
        return
    _resolved = True

    try:
        result = subprocess.run(
            ["git", "rev-parse", "--show-toplevel"],
            capture_output=True,
            text=True,
            timeout=3,
        )
        if result.returncode == 0 and result.stdout:
            _cached_project = os.path.basename(result.stdout.strip())
    except Exception:
        pass

    try:
        result = subprocess.run(
            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
            capture_output=True,
            text=True,
            timeout=3,
        )
        if result.returncode == 0 and result.stdout:
            _cached_branch = result.stdout.strip()
    except Exception:
        pass


def infer_git_project() -> str | None:
    env = os.environ.get("RELOG_PROJECT")
    if env:
        return env
    _resolve_git()
    return _cached_project


def infer_git_branch() -> str | None:
    env = os.environ.get("RELOG_BRANCH")
    if env:
        return env
    _resolve_git()
    return _cached_branch
